sdk/src/Core/Amazon.Runtime/Pipeline/HttpHandler/_bcl/HttpWebRequestFactory.cs

```diff
@@ -474,6 +474,14 @@ public virtual void ConfigureRequest(IRequestContext requestContext)
                 requestContext.Metrics.AddProperty(Metric.ProxyPort, requestContext.ClientConfig.ProxyPort);
                 _request.Proxy = proxy;
             }
+            else if (_request.RequestUri.Scheme == Uri.UriSchemeHttp)
+            {
```

Comment:
AWS service endpoints are usually `HTTPS` (I took a quick look at https://docs.aws.amazon.com/general/latest/gr/aws-service-information.html and a lot of services don't even have an `HTTP` option).

With this approach of checking the scheme, wouldn't it be possible for a customer to set the `http_proxy` variable, which then it'd be ignored when trying to connect to `https://<service>.<region>.amazonaws.com`?

Reply:
Hey! Yeah this is a point of confusion, and poor cross-sdk behavior I've run into. The original idea behind these two environment variable (as far as I can tell from the community at large) is that the scheme of what you want to connect too matters. And as for AWS you're right `http_proxy` would basically never be used (unless you're maybe talking to like dynamodb local or something similar).

You could in theory still connect to a proxy being served over HTTPS when trying to reach an HTTP URL, and you can in theory (and may want too) connect to an Proxy over plain HTTP when trying to connect to a HTTPS site.

  - See wget mentioning which should be used depending on the connection type *not the proxies connection type): https://www.gnu.org/software/wget/manual/html_node/Proxies.html
  - Curl's environment variables: https://everything.curl.dev/usingcurl/proxies/env
  - Java does the same with properties/variables: https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html

The reason you'd want a distinction between the two is because proxying to an HTTPS URL requires handling http `CONNECT` which forces a new connection, and is in general not the most widely supported things (it also means you're probably able to do less actual rewriting of traffic unless you have a cert on the clients machine or what have you, so you may have different behavior).

---

That being said some SDKs in AWS do what this PR does where the environment variable is properly based off of the scheme, some of them fallback to http proxy even on https urls, and some don't support it. I'm going for what is probably most universally "correct" here, and what _most_ of the SDKs do, connect depending on the scheme of the request. If you want me to fallback though like some of the other SDKs do I can change this behavior.

Reply:
Got it, thanks a lot for the explanation!

I understand it's frustrating that the SDKs are not consistent at the moment, but if we had to pick a behavior to "copy", I'd go with following what the CLI does (it was the original request and it's a very common setup for our customers to have both the .NET SDK and CLI in their environments).

Reply:
No problem at all! So the CLI is doing what we're doing here. It uses `http_proxy` for HTTP scheme requests, and `https_proxy` for HTTPS scheme requests. There is no fallback within the CLI. So we should be good to go!